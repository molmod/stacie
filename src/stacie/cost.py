# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024-2025 The contributors of the Stacie Python Package.
# See the CONTRIBUTORS.md file in the project root for a full list of contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Cost function to optimize models for the low-frequency part of the spectrum."""

import attrs
import numpy as np
from numpy.typing import ArrayLike, NDArray
from scipy.special import digamma, gammaln

from .model import SpectrumModel

__all__ = ("LowFreqCost", "entropy_gamma", "logpdf_gamma")


@attrs.define
class LowFreqCost:
    """Cost function to fit a model to the low-frequency part of the spectrum."""

    freqs: NDArray[float] = attrs.field()
    """The frequencies for which the spectrum amplitudes are computed."""

    ndofs: NDArray[int] = attrs.field()
    """The number of independent contributions to each spectrum amplitude."""

    amplitudes: NDArray[float] = attrs.field()
    """The actual spectrum amplitudes at frequencies in ``self.freqs``."""

    weights: NDArray[float] = attrs.field()
    """The fitting weights for each grid point."""

    model: SpectrumModel = attrs.field()
    """The model to be fitted to the spectrum."""

    def __call__(self, pars: ArrayLike, deriv: int = 0) -> float:
        """Evaluate the cost function and its derivatives.

        Parameters
        ----------
        pars
            The parameter vector for which the loss function must be computed.
        deriv
            The order of derivatives of the cost function to include.

        Returns
        -------
        results
            A list with the cost function and the requested derivatives.
        """
        # Prepare result arrays, with inf and nan by default.
        # These will be overwritten if the model is valid.
        pars = np.asarray(pars)
        vec_shape = pars.shape[:-1]
        par_shape = pars.shape[-1:]
        results = [np.full(vec_shape, np.inf)]
        if deriv >= 1:
            results.append(np.full(vec_shape + par_shape, np.nan))
        if deriv >= 2:
            results.append(np.full(vec_shape + par_shape + par_shape, np.nan))
        if deriv >= 3:
            raise ValueError("Third or higher derivatives are not supported.")

        mask = self.model.valid(pars)
        if not mask.any():
            return results

        # Compute the model spectrum and its derivatives.
        amplitudes_model = self.model.compute(
            self.freqs, pars if mask.ndim == 0 else pars[mask], deriv
        )
        kappas = 0.5 * self.ndofs

        # Only continue with parameters for which the model does not become negative.
        pos_mask = (amplitudes_model[0] > 0).all(axis=-1)
        if not pos_mask.any():
            return results
        if mask.ndim > 0:
            amplitudes_model = [am[pos_mask] for am in amplitudes_model]
            mask[mask] = pos_mask
        del pos_mask

        # Log-likelihood computed with the scaled Chi-squared distribution.
        # The Gamma distribution is used because the scale parameter is easily incorporated.
        thetas = [am / kappas for am in amplitudes_model]
        ll_terms = logpdf_gamma(self.amplitudes, kappas, thetas[0], deriv)
        results[0][mask] = -np.einsum("...i,i->...", ll_terms[0], self.weights)
        if deriv >= 1:
            results[1][mask] = -np.einsum(
                "...pi,...i,i->...p", thetas[1], ll_terms[1], self.weights
            )
        if deriv >= 2:
            results[2] = -(
                np.einsum(
                    "...pi,...qi,...i,i->...pq", thetas[1], thetas[1], ll_terms[2], self.weights
                )
                + np.einsum("...pqi,...i,i->...pq", thetas[2], ll_terms[1], self.weights)
            )
        return results


def logpdf_gamma(x: NDArray[float], kappa: NDArray[float], theta: NDArray[float], deriv: int = 1):
    """Compute the logarithm of the probability density function of the Gamma distribution.

    Parameters
    ----------
    x
        The argument of the PDF (random variable).
        Array with shape ``(nfreq,)``.
    kappa
        The shape parameter.
        Array with shape ``(nfreq,)``.
    theta
        The scale parameter.
        Array with shape ``(..., nfreq,)``.
    deriv
        The order of the derivatives toward theta to compute: 0, 1 or 2.

    Returns
    -------
    results
        A list of results (function value and requested derivatives.)
        All elements have the same shape as the ``theta`` array.
    """
    kappa = np.asarray(kappa)
    theta = np.asarray(theta)
    ratio = np.asarray(x) / theta
    results = [-gammaln(kappa) - np.log(theta) + (kappa - 1) * np.log(ratio) - ratio]
    if deriv >= 1:
        results.append((ratio - kappa) / theta)
    if deriv >= 2:
        results.append((kappa - 2 * ratio) / theta**2)
    if deriv >= 3:
        raise ValueError("Third or higher derivatives are not supported.")
    return results


def entropy_gamma(kappa: NDArray[float], theta: NDArray[float], deriv: int = 1):
    """Compute the entropy of the Gamma distribution.

    Parameters
    ----------
    kappa
        The shape parameter.
    theta
        The scale parameter.
    deriv
        The order of the derivatives toward theta to compute: 0, 1 or 2.

    Returns
    -------
    results
        A list of results (function value and requested derivatives.)
        All elements have the same shape as the ``kappa`` and ``theta`` arrays.
    """
    kappa = np.asarray(kappa)
    theta = np.asarray(theta)
    results = [kappa + np.log(theta) + gammaln(kappa) + (1 - kappa) * digamma(kappa)]
    if deriv >= 1:
        results.append(1 / theta)
    if deriv >= 2:
        results.append(-1 / theta**2)
    if deriv >= 3:
        raise ValueError("Third or higher derivatives are not supported.")
    return results
