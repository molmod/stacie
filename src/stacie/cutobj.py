# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Objective functions used to decide which part of the spectrum to fit to."""

from collections.abc import Callable
from typing import NewType

import jax
import jax.numpy as jnp
from jax.typing import ArrayLike as JArrayLike

__all__ = ("CutObj", "cutobj_symcu")


CutObj = NewType("CutObj", Callable[[JArrayLike, JArrayLike, JArrayLike], jax.Array])


def cutobj_symcu(amplitudes: JArrayLike, kappas: JArrayLike, thetas: JArrayLike) -> jax.Array:
    """Compute the excess variance of the cumulative sum of the normal errors, starting from middle.

    Parameters
    ----------
    amplitudes
        Spectrum amplitudes.
    kappas
        The Gamma shape parameters.
    thetas
        The Gamma scale parameters.

    Returns
    -------
    obj
        The objective to minimize.
    """
    # Transformation to normal errors
    uni = jax.scipy.stats.gamma.cdf(amplitudes, kappas, scale=thetas)
    nor = jax.scipy.special.erfinv(2 * uni - 1) * jnp.sqrt(2)
    # Objective = minimize excess variance of the cumulative sum of the normal noise.
    return ((jnp.cumsum(nor) - jnp.sum(nor) / 2) ** 2).mean() - len(nor) / 4
