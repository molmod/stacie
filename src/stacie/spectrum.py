# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Utility to prepare spectrum and other inputs for given time series."""

import attrs
import numpy as np
from numpy.typing import NDArray

__all__ = ("Spectrum", "prepare_acfint")


@attrs.define
class Spectrum:
    """Container class holding all the inputs for the ACF integral esstimate."""

    ndofs: NDArray[int] = attrs.field()
    """The number of independnt contributions to each amplitude."""

    prefactor: float = attrs.field()
    """The given prefactor for the spectrum to get the right units for the ACF integral."""

    times: NDArray[float] = attrs.field()
    """The equidistant time axis of the sequences, always starts at zero."""

    freqs: NDArray[float] = attrs.field()
    """The equidistant frequency axis of the spectrum."""

    amplitudes: NDArray[float] = attrs.field()
    """The spectrum amplitudes averaged over the given input sequences."""

    amplitudes_ref: NDArray[float] | None = attrs.field(default=None)
    """Optionally, the known analytical model of the power spectrum, on the same frequency grid."""

    @property
    def nstep(self) -> int:
        """The number of time steps of the input sequences."""
        return len(self.times)

    @property
    def timestep(self) -> float:
        """The time span between two subsequent items in the given time series."""
        return self.times[1] - self.times[0]

    @property
    def nfreq(self) -> int:
        """The number of irfft frequency grid points."""
        return len(self.freq)


def prepare_acfint(
    sequences: NDArray[float], *, prefactor: float = 0.5, timestep: float = 1
) -> Spectrum:
    """Compute a spectrum and store all physical inputs in a ``Spectrum`` instance.

    Parameters
    ----------
    sequences
        The input sequences, array with shape `(nindep, nstep)`,
        of which each row is a time-dependent sequence.
        All sequences are assumed to be statistically independent.
        The time step is assumed to be 1.
        If needed, multiply the resulting acfint and acfint_err with the appropriate time step.
    prefactor
        A factor to be multiplied with the autocorrelation function
        to give it a physically meaningful unit.
    timestep
        The timestep of the input sequence.

    Returns
    -------
    spectrum
        An instance of the ``Spectrum`` object holding all the inputs needed to estimate
        the integral of the autocorrelation function.
    """
    # Get basic parameters of the input sequences.
    nindep, nstep = sequences.shape
    times = np.arange(nstep) * timestep
    freqs = np.fft.rfftfreq(nstep, d=timestep)

    # Number of "degrees of freedom" (contributions) to each amplitude
    ndofs = np.full(freqs.shape, 2 * nindep)
    ndofs[0] = nindep
    if len(freqs) % 2 == 0:
        ndofs[-1] = nindep

    # Compute the spectrum and scale it.
    amplitudes = (abs(np.fft.rfft(sequences, axis=1)) ** 2).mean(axis=0)
    amplitudes *= prefactor * timestep / nstep

    return Spectrum(ndofs, prefactor, times, freqs, amplitudes)
