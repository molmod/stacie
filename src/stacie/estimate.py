# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Algorithm to estimate the autocorrelation integral."""

import attrs
import numpy as np
from numpy.typing import NDArray
from scipy.optimize import minimize

from .model import LowFreqCost
from .rpi import rpi_opt
from .spectrum import Spectrum

__all__ = ("Result", "HessianError", "estimate_acfint", "fit_model_spectrum")


@attrs.define
class Result:
    """Container class holding all the results of the ACF integral estimate."""

    spectrum: Spectrum = attrs.field()
    """The input spectrum from which the ACF integral is estimated."""

    ncut: int = attrs.field()
    """The low-pass cutoff index of the frequency axis, to suppress periodic boundary artifacts."""

    history: dict[int, dict[str]] = attrs.field()
    """History of ncut optimization."""

    @property
    def props(self) -> dict[str]:
        """Properties computed from the fit up to the selected spectrum cutoff."""
        return self.history[self.ncut]

    @property
    def acfint(self) -> float:
        """Integral of the autocorrelation integral."""
        return self.props["pars"][:2].sum()

    @property
    def acfint_var(self) -> float:
        """Variance of the  of the autocorrelation integral."""
        return self.props["covar"][:2, :2].sum()

    @property
    def acfint_std(self) -> float:
        """Uncertainty of the autocorrelation integral."""
        return np.sqrt(self.acfint_var)

    @property
    def corrtime_tail(self) -> float:
        """Slowest time scale in the signal."""
        return self.props["pars"][2]

    @property
    def corrtime_tail_var(self) -> float:
        """Variance of the slowest time scale in the signal."""
        return self.props["covar"][2, 2]

    @property
    def corrtime_tail_std(self) -> float:
        """Uncertianty of the slowest time scale in the signal."""
        return np.sqrt(self.corrtime_tail_var)


class HessianError(ValueError):
    """Raised when the Hessian is infinite or not strictly positive definite."""


def estimate_acfint(
    spectrum: Spectrum,
    *,
    fcut: float | None = None,
    maxscan: int = 100,
    ncutmin: int = 10,
) -> Result:
    """Estimate the integral of the autocorrelation function.

    It is recommended to leave the keyword arguments to their default values,
    except for methodological testing.

    Parameters
    ----------
    spectrum
        A ``Spectrum`` instance holding all the inputs for the estimation of the ACF integral.
    fcut
        The highest value of the spectrum to consider
    maxscan
        The maximum number of cutoffs to test.
        If 1, then only the given fcut is used.
    ncutmin
        The minimal amount of frequency data points to use in the fit.

    Returns
    -------
    result
        A ``Result`` instance with inputs, intermediate results and outputs.
    """
    history = {}

    def objective(n: int, check_hess: bool = False):
        ncut = ncutmin + n * stride
        props = fit_model_spectrum(
            spectrum.timestep,
            spectrum.freqs[:ncut],
            spectrum.amplitudes[:ncut],
            spectrum.ndofs[:ncut],
        )
        if check_hess:
            # When scanning, only good minima are saved and treated as valid
            hess_evals = props["hess_evals"]
            if not (np.isfinite(hess_evals).all() and (hess_evals > 0).all()):
                return np.inf
        history[ncut] = props
        return props["obj"]

    ncutmax = len(spectrum.freqs) if fcut is None else spectrum.freqs.searchsorted(fcut)
    if ncutmax < ncutmin:
        raise ValueError("Too few data points for fit.")
    if maxscan == 1:
        ncut = ncutmax
        stride = 1
        if not np.isfinite(objective(ncutmax - ncutmin)):
            raise ValueError("Fit failed")
    else:
        stride = max((ncutmax - ncutmin) // maxscan, 1)
        rpi_opt(objective, [0, maxscan], mode="min")
        if len(history) == 0:
            raise AssertionError("Could not find a suitable solution")
        ncut = min((record["obj"], key) for key, record in history.items())[1]

    return Result(spectrum, ncut, dict(sorted(history.items())))


def fit_model_spectrum(
    timestep: float, freqs: NDArray[float], amplitudes: NDArray[float], ndofs: NDArray[int]
) -> dict[str, NDArray]:
    # Maximize likelihood
    cost = LowFreqCost(timestep, freqs, amplitudes, ndofs)
    pars_init = cost.guess
    if not cost.valid(pars_init):
        raise AssertionError("Infeasible guess")
    opt = minimize(
        cost.func,
        pars_init,
        jac=cost.grad,
        hess=cost.hess,
        bounds=cost.bounds,
        method="trust-constr",
    )

    # Compute all properties
    hess = cost.hess(opt.x)
    evals, evecs = np.linalg.eigh(hess)
    props = cost.prop(opt.x)
    props["hess"] = hess
    props["hess_evals"] = evals
    props["hess_evecs"] = evecs
    props["covar"] = np.linalg.inv(hess)
    return props
