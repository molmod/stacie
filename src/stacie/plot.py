# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Plot the results of the estimate of the ACF integral."""

import matplotlib.pyplot as plt
import numpy as np
from matplotlib.backends.backend_pdf import PdfPages
from scipy import stats

from .estimate import Result

__all__ = ("plot",)


def plot(path_pdf: str, res: Result | list[Result], acf_unit: str = "A", time_unit: str = "s"):
    """Plot the results of the ACF integral estimation.

    Parameters
    ----------
    path_pdf
        The PDF file where all the figures are stored.
    res
        A single Results instance or a list of them.
        If the (first) result instance has model_spectrum set,
        theoretical expectations are included.
        When multiple results instances are given,
        only the first one is plotted in blue.
        All remaining ones are plotted in light grey.
    acf_unit
        A string with the unit of the autocorrelation function times the prefactor
    time_unit
        A string with the unit of the time axis, used for labels.
    """
    # Prepare res
    if isinstance(res, Result):
        res = [res]
    sfac = 2

    model_props = {"ls": "--", "color": "k", "alpha": 0.5}

    def plot_spectrum(ax, r: Result):
        nplot = 2 * r.ncut
        s = r.spectrum
        ax.plot(s.freqs[:nplot], s.amplitudes[:nplot], color="C0", lw=1)
        mean = r.props["spectrum_model"]
        freqs = s.freqs[: len(mean)]
        ax.plot(freqs, mean, color="C2")
        # ax.fill_between(freqs, mean - sfac * std, mean + sfac * std, color="C2", alpha=0.3, lw=0)
        ax.axvline(s.freqs[r.ncut - 1], ymax=0.1, color="k")
        if s.amplitudes_ref is not None:
            ax.plot(s.freqs[:nplot], s.amplitudes_ref[:nplot], **model_props)
        ax.set_xlabel(f"Frequency [1/{time_unit}]")
        ax.set_ylabel(f"Spectrum [{acf_unit} {time_unit}]")
        ax.set_title(f"ACF integral = {r.acfint:.5f} Â± {r.acfint_std:.5f}")

    def plot_all_models(ax, r):
        s = r.spectrum
        for ncut, props in r.history.items():
            mean = props["spectrum_model"]
            freqs = s.freqs[: len(mean)]
            if ncut == r.ncut:
                ax.plot(freqs, mean, color="k", lw=2, zorder=2.5)
            else:
                ax.plot(freqs, mean, color="C2", lw=1, alpha=0.5)
        if s.amplitudes_ref is not None:
            nplot = min(2 * max(r.history), len(s.freqs))
            ax.plot(s.freqs[:nplot], s.amplitudes_ref[:nplot], **model_props)
        ax.set_xlabel(f"Frequency [1/{time_unit}]")
        ax.set_ylabel(f"Model Spectrum [{acf_unit} {time_unit}]")

    def plot_objective(ax, r):
        freqs = []
        objs = []
        s = r.spectrum
        for ncut, props in sorted(r.history.items()):
            freqs.append(s.freqs[ncut])
            objs.append(props["obj"])

        ax.plot(freqs, objs, color="C1", lw=1)
        ax.plot([s.freqs[r.ncut]], [r.props["obj"]], marker="o", color="k", ms=2)
        ax.set_xlabel(f"Cutoff frequency [1/{time_unit}]")
        ax.set_ylabel("Objective function")
        ncutmax = max(r.history)
        ax.set_ylim(r.props["obj"] * 1.2, ncutmax / 4)

    def plot_uncertainty(ax, r):
        freqs = []
        acfints = []
        acfint_stds = []
        s = r.spectrum
        for ncut, props in sorted(r.history.items()):
            freqs.append(s.freqs[ncut])
            acfints.append(props["pars"][:2].sum())
            acfint_stds.append(np.sqrt(props["covar"][:2, :2].sum()))
        acfints = np.array(acfints)
        acfint_stds = np.array(acfint_stds)
        ax.plot(freqs, acfints, "C3")
        ax.fill_between(
            freqs,
            acfints - sfac * acfint_stds,
            acfints + sfac * acfint_stds,
            color="C3",
            alpha=0.3,
            lw=0,
        )
        s = r.spectrum
        ax.errorbar(
            [s.freqs[r.ncut]], [r.acfint], [r.acfint_std * sfac], marker="o", ms=2, color="k"
        )
        if s.amplitudes_ref is not None:
            limit = s.amplitudes_ref[0]
            ax.axhline(limit, **model_props)
        ax.set_xlabel(f"Cutoff frequency [1/{time_unit}]")
        ax.set_ylabel(f"ACF integral [{acf_unit} {time_unit}]")

    def plot_evals(ax, r):
        freqs = []
        evals = []
        s = r.spectrum
        for ncut, props in sorted(r.history.items()):
            freqs.append(s.freqs[ncut])
            evals.append(np.linalg.eigvalsh(props["covar"]))
            if ncut == r.ncut:
                ax.plot([freqs[-1]], [evals[-1]], color="k", marker="o", ms=2, zorder=2.5)
        evals = np.array(evals)
        ax.plot(freqs, evals, color="C4")
        ax.set_xlabel(f"Cutoff frequency [1/{time_unit}]")
        ax.set_ylabel("Covariance eigenvalues")
        ax.set_yscale("log")

    def plot_uni(ax, r):
        nor = r.props["nor"]
        ax.plot((np.cumsum(nor) - nor.sum() / 2) ** 2)
        # spec = 2 * abs(np.fft.rfft(nor)[1:])**2 / len(nor)
        # ax.set_title(format(stats.chi2(2).logpdf(spec).mean(), ".3e"))
        # ax.plot(spec)
        # ax.plot(spec.real)
        # ax.plot(spec.imag)
        # print(nor.std())
        # print(spec.real.std())
        # print(spec.imag[1:].std())
        # print()
        # qgrid = (np.arange(len(uni)) + 0.5) / len(uni)
        # ax.plot(qgrid, uni)

    def plot_qq(ax):
        r0 = res[0]
        cdfs = (np.arange(len(res)) + 0.5) / len(res)
        quantiles = stats.norm().ppf(cdfs)
        limit = r0.spectrum.amplitudes_ref[0]
        normed_errors = np.array([(r.acfint - limit) / r.acfint_std for r in res])
        normed_errors.sort()
        ax.scatter(quantiles, normed_errors, c="C0", s=3)
        ax.plot([-2, 2], [-2, 2], **model_props)
        ax.set_xlabel("Normal quantiles [1]")
        ax.set_ylabel("Sorted normalized errors [1]")
        ax.set_title("QQ Plot")

    def plot_acfint_estimates(ax):
        values = np.array([r.acfint for r in res])
        errors = np.array([sfac * r.acfint_std for r in res])
        order = values.argsort()
        values = values[order]
        errors = errors[order]
        ax.errorbar(np.arange(len(res)), values, errors, fmt="o", lw=1, ms=2, ls="none")
        r0 = res[0]
        if r0.spectrum.amplitudes_ref is not None:
            limit = r0.spectrum.amplitudes_ref[0]
            ax.axhline(limit, **model_props)
        ax.set_xlabel("Rank")
        ax.set_ylabel(f"Mean and uncertainties [{acf_unit} {time_unit}]")
        ax.set_title("ACF Integral")

    with PdfPages(path_pdf) as pdf:
        for r in res:
            fig, ax = plt.subplots(figsize=(6, 6))
            plot_spectrum(ax, r)
            pdf.savefig(fig)
            plt.close(fig)

            if len(r.history) > 1:
                fig, axs = plt.subplots(2, 2, figsize=(6, 6))
                plot_all_models(axs[0, 0], r)
                plot_objective(axs[0, 1], r)
                plot_uncertainty(axs[1, 0], r)
                # plot_evals(axs[1, 1], r)
                plot_uni(axs[1, 1], r)
                pdf.savefig(fig)
                plt.close(fig)

        if len(res) > 1:
            if res[0].spectrum.amplitudes_ref is not None:
                fig, ax = plt.subplots(figsize=(6, 6))
                plot_qq(ax)
                pdf.savefig(fig)
                plt.close(fig)

            fig, ax = plt.subplots(figsize=(6, 6))
            plot_acfint_estimates(ax)
            pdf.savefig(fig)
            plt.close(fig)
