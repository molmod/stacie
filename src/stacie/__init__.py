# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024-2025 The contributors of the Stacie Python Package.
# See the CONTRIBUTORS.md file in the project root for a full list of contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""The Stacie package."""

from __future__ import annotations

from .estimate import estimate_acint
from .plot import UnitConfig, plot_results
from .spectrum import compute_spectrum
from .summary import summarize_results

__all__ = (
    "UnitConfig",
    "__version__",
    "__version_tuple__",
    "compute_spectrum",
    "estimate_acint",
    "plot_results",
    "summarize_results",
)


try:
    from ._version import __version__, __version_tuple__
except ImportError:
    __version__ = "0.0.0a-dev"
    __version_tuple__ = (0, 0, 0, "a-dev")
