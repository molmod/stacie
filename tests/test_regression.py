# Stacie is a STable AutoCorrelation Integral Estimator.
# Copyright (C) 2024  Toon Verstraelen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --
"""Regression tests for typical Stacie workflows."""

import numpy as np
import pytest
from path import Path
from stacie.estimate import estimate_acfint
from stacie.plot import plot
from stacie.spectrum import Spectrum


@pytest.fixture()
def synth_30():
    amplitudes = np.load("tests/spectrum_30.npy")
    amplitudes_ref = np.load("tests/spectrum_ref.npy")
    ndof = np.full(amplitudes.shape, 800)
    ndof[0] = 400
    ndof[-1] = 400
    nstep = 2000
    freqs = np.fft.rfftfreq(nstep)
    return Spectrum(ndof, 0.5, np.arange(nstep), freqs, amplitudes, amplitudes_ref)


@pytest.fixture()
def synth_36():
    amplitudes = np.load("tests/spectrum_36.npy")
    amplitudes_ref = np.load("tests/spectrum_ref.npy")
    ndof = np.full(amplitudes.shape, 800)
    ndof[0] = 400
    ndof[-1] = 400
    nstep = 2000
    freqs = np.fft.rfftfreq(nstep)
    return Spectrum(ndof, 0.5, np.arange(nstep), freqs, amplitudes, amplitudes_ref)


def plot_test_result(prefix, res):
    dn_out = Path("tests/outputs")
    dn_out.makedirs_p()
    path_pdf = dn_out / f"{prefix}.pdf"
    plot(path_pdf, res)


def check_noscan_single(spectrum, regtest, prefix):
    res = estimate_acfint(spectrum, fcut=0.005, maxscan=1)
    with regtest:
        print(f"ACF Int  = {res.acfint:.4f} ± {res.acfint_std:.4f}")
        print(f"Tau tail = {res.corrtime_tail:.4f} ± {res.corrtime_tail_std:.4f}")
        print(f"Log lh   = {res.props['ll']:.5e}")
    plot_test_result(prefix, res)


def test_noscan_30(synth_30, regtest):
    check_noscan_single(synth_30, regtest, "noscan_30")


def test_noscan_36(synth_36, regtest):
    check_noscan_single(synth_36, regtest, "noscan_36")


def test_noscan_multi(synth_30, synth_36, regtest):
    res = []
    for spectrum in synth_30, synth_36:
        r = estimate_acfint(spectrum, fcut=0.005, maxscan=1)
        with regtest:
            print(f"ACF Int  = {r.acfint:.4f} ± {r.acfint_std:.4f}")
            print(f"Tau tail = {r.corrtime_tail:.4f} ± {r.corrtime_tail_std:.4f}")
            print(f"Log lh   = {r.props['ll']:.5e}")
            print("--")
        res.append(r)
    plot_test_result("noscan_multi", res)


def check_scan_single(spectrum, regtest, prefix):
    res = estimate_acfint(spectrum, fcut=0.01, maxscan=10)
    with regtest:
        print(f"ACF Int  = {res.acfint:.4f} ± {res.acfint_std:.4f}")
        print(f"Tau tail = {res.corrtime_tail:.4f} ± {res.corrtime_tail_std:.4f}")
        print(f"Log lh   = {res.props['ll']:.5e}")
    plot_test_result(prefix, res)


def test_scan_30(synth_30, regtest):
    check_scan_single(synth_30, regtest, "scan_30")


def test_scan_36(synth_36, regtest):
    check_scan_single(synth_36, regtest, "scan_36")


def test_scan_multi(synth_30, synth_36, regtest):
    res = []
    for spectrum in synth_30, synth_36:
        r = estimate_acfint(spectrum, fcut=0.01, maxscan=10)
        with regtest:
            print(f"ACF Int  = {r.acfint:.4f} ± {r.acfint_std:.4f}")
            print(f"Tau tail = {r.corrtime_tail:.4f} ± {r.corrtime_tail_std:.4f}")
            print(f"Log lh   = {r.props['ll']:.5e}")
            print("--")
        res.append(r)
    plot_test_result("scan_multi", res)
