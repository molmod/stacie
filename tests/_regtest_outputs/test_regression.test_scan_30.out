ACF Int  = 148.7408 ± 6.3109
Tau tail = 185.0730 ± 20.8904
Log lh   = -2.98082e+01
